src/cls/cmpomap/types.h

```diff
@@ -0,0 +1,70 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2020 Red Hat, Inc
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ */
+
+#pragma once
+
+#include <string>
+#include <boost/container/flat_map.hpp>
+#include "include/rados.h" // CEPH_OSD_CMPXATTR_*
+#include "include/encoding.h"
+
+namespace cls::cmpomap {
+
+/// comparison operand type
+enum class Mode : uint8_t {
+  String = CEPH_OSD_CMPXATTR_MODE_STRING,
```

Comment:
are these the only 2 possible types in xattrs?

Reply:
yeah. you can find these in include/rados.h

Reply:
so why do you repeat these enums here?
maybe better to just take from there, and if someone change them there it will (hopefully) break here?
(e.g. switch statement no covering all cases)

Reply:
yeah good question. one of the things that _really_ bothered me about the librados `omap_cmp()` op is that it used these `CEPH_OSD_CMPXATTR_` enums directly, but didn't actually support U64 types or any string comparisons except EQ/LT/GT

also, these are unnamed enums in rados.h, so we'd have to store their values as a raw integer. that means the compiler can't tell whether we're missing switch cases

so i thought it was much better to explicitly spell out the supported modes and ops with 'enum class' in the cls::cmpomap namespace. if new types or ops are added for cmpxattr, they may or may not make sense for omap - so i'd prefer to keep a degree of separation